### DISCUSSION TITLE
v2.0.0-rc.2

### DISCUSSION THREAD

**Discussion by ruslandoga** [plausible/analytics#3011] [Self-Hosted Support]
The highlights of this release are:

- more visitor metrics like total visits and views per visit
- period comparisons [[docs]](https://plausible.io/docs/guided-tour#compare-your-stats-over-time) [[example]](https://plausible.io/plausible.io?comparison=previous_period)
- ability to change domains names https://github.com/plausible/analytics/pull/2803
- automatic geolocation database downloads  plausible/analytics#2362
- more email adapters support (Mailgun, Mandrill, SendGrid) plausible/analytics#2649

### Upgrading Plausible Analytics to v2.0

>  ⚠️ ⚠️ ⚠️
> 
> Upgrading to `v2.0` requires performing a data migration.
> Please read these notes until the end before deploying `v2.0.0-rc.2`

#### Ensure you are using a new ClickHouse version

The steps below have been tested with [`clickhouse/clickhouse-server:22.6-alpine`](https://github.com/plausible/hosting/blob/3e1462eabd510cc9320a6cfed0ae45355d86bb05/docker-compose.yml#L16-L26) please make sure to upgrade ClickHouse to at least this version.

<details>
<summary>Here's the excerpt from <a href="https://github.com/plausible/analytics/discussions/2507"><code>v1.5 release discussion</code></a> regarding the ClickHouse upgrade</summary>

---

In your docker-compose.yml update the image used for plausible_events_db to a newer ClickHouse version:

### Upgrading ClickHouse to 22.6

```diff
plausible_events_db:
- image: yandex/clickhouse-server:21.3.2.5
+ image: clickhouse/clickhouse-server:22.6-alpine
```

#### Restart the container

```console
$ docker compose stop plausible plausible_events_db
$ docker compose rm plausible_events_db
$ docker compose up -d
```

This will boot up the new version of ClickHouse.

Related PR: https://github.com/plausible/hosting/pull/45

---

</details>

#### Update image tag

In your `docker-compose.yml` update the image used for `plausible` to `v2.0.0-rc.2`

```diff
plausible:
- image: plausible/analytics:v1.5.1
+ image: plausible/analytics:v2.0.0-rc.2
```

and restart the container

```console
$ cd hosting
$ docker compose stop plausible
$ docker compose rm plausible
$ docker compose up -d
```

This will boot up the new version of the app.

If you open the dashboards now, you wouldn't see any past metrics. This is expected as `v2.0` uses the new `events_v2` and `sessions_v2` tables to store analytics data. We need to perform data migration to copy the data into the new tables.

#### Run data migration

> ⚠️ ⚠️ ⚠️
> 
> You need enough free disk space available for x2 of the current `plausible_events_db`'s `event-data` volume size. You can use something like `docker system df -v | grep hosting_event-data` to check how much space the current volume is occupying.

Connect to the running `plausible` container and start the migration flow

```console
$ cd hosting
$ docker compose exec plausible bin/plausible rpc Plausible.DataMigration.NumericIDs.run
```

You can attempt this migration multiple times unless you drop `v1` tables.

#### Drop v1 tables (optional)

Once you **verify** the migration went **well,** the old tables can be dropped. It's easiest to use `clickhouse-client` for this

```console
$ ch hosting
$ docker compose exec plausible_events_db clickhouse-client
```
```sql
:) \c plausible_events_db
:) show tables;
:) drop table events;
:) drop table sessions;
```

See https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#max-table-size-to-drop for how to drop tables with more than 50GB of data.

#### Enable automatic MaxMind GeoLite2 updates (optional)

In your `plausible-config.env` set [`MAXMIND_LICENSE_KEY`](https://www.maxmind.com/en/accounts/current/license-key) environment variable and get an automatically updated [GeoLite2 City](https://dev.maxmind.com/geoip/geolite2-free-geolocation-data) geolocation database. The database edition is configurable with `MAXMIND_EDITION` environment variable and defaults to `GeoLite2-City`.

```diff
- IP_GEOLOCATION_DB=...
+ MAXMIND_LICENSE_KEY=LNpsJCCKPis6XvBP
+ MAXMIND_EDITION=GeoLite2-City
```

Note that for the changes in `plausible-config.env` to propagate to the `plausible`, the container needs to be recreated:

```console
$ docker compose stop plausible
$ docker compose rm plausible
$ docker compose up -d
```

Also note that using `GeoLite2-City` edition requires more RAM than using `GeoLite2-Country`.

Now you can remove any other volumes and services used to download, store, and update geolocation databases.

## Changelog

Following [changes](https://github.com/plausible/analytics/compare/v1.5.1...v2.0.0-rc.2) have been made since v1.5:

### Added
- Add support for `with_imported=true` in Stats API aggregate endpoint
- Ability to use '--' instead of '=' sign in the `tagged-events` classnames
- 'Last updated X seconds ago' info to 'current visitors' tooltips
- Add support for more Bamboo adapters, i.e. `Bamboo.MailgunAdapter`, `Bamboo.MandrillAdapter`, `Bamboo.SendGridAdapter` plausible/analytics#2649
- Ability to change domain for existing site (requires numeric IDs data migration, instructions will be provided separately) UI + API (`PUT /api/v1/sites`)
- Add `LOG_FAILED_LOGIN_ATTEMPTS` environment variable to enable failed login attempts logs plausible/analytics#2936
- Add `MAILER_NAME` environment variable support plausible/analytics#2937
- Add `MAILGUN_BASE_URI` support for `Bamboo.MailgunAdapter` plausible/analytics#2935
- Add a landing page for self-hosters plausible/analytics#2989

### Fixed
- Fix tracker bug - call callback function even when event is ignored
- Make goal-filtered CSV export return only unique_conversions timeseries in the 'visitors.csv' file
- Stop treating page filter as an entry page filter
- City report showing N/A instead of city names with imported data plausible/analytics#2675
- Empty values for Screen Size, OS and Browser are uniformly replaced with "(not set)"
- Fix [more pageviews with session prop filter than with no filters](https://github.com/plausible/analytics/issues/1666)
- Cascade delete sent_renewal_notifications table when user is deleted plausible/analytics#2549
- Show appropriate top-stat metric labels on the realtime dashboard when filtering by a goal
- Fix breakdown API pagination when using event metrics plausible/analytics#2562
- Automatically update all visible dashboard reports in the realtime view
- Connect via TLS when using HTTPS scheme in ClickHouse URL plausible/analytics#2570
- Add error message in case a transfer to an invited (but not joined) user is requested plausible/analytics#2651
- Fix bug with [showing property breakdown with a prop filter](https://github.com/plausible/analytics/issues/1789)
- Fix bug when combining goal and prop filters plausible/analytics#2654
- Fix broken favicons when domain includes a slash

### Changed
- Treat page filter as entry page filter for `bounce_rate`
- Reject events with long URIs and data URIs plausible/analytics#2536
- Always show direct traffic in sources reports plausible/analytics#2531
- Stop recording XX and T1 country codes plausible/analytics#2556
- Device type is now determined from the User-Agent instead of window.innerWidth plausible/analytics#2711
- Add padding by default to embedded dashboards so that shadows are not cut off plausible/analytics#2744
- Update the User Agents database (https://github.com/matomo-org/device-detector/releases/tag/6.1.1)

### Removed
- Remove Firewall plug and `IP_BLOCKLIST` environment variable
- Remove the ability to collapse the main graph plausible/analytics#2627
- Remove `custom_dimension_filter` feature flag plausible/analytics#2996

<hr /><em>This discussion was created from the release <a href='https://github.com/plausible/analytics/releases/tag/v2.0.0-rc.2'>v2.0.0-rc.2</a>.</em>

**Comment by Olakeen** [comment]
Still + image: plausible/analytics:v2.0.0-rc.1 ? not + image: plausible/analytics:v2.0.0-rc.2 ?

**Reply by ruslandoga** [reply]
👋 @Olakeen

Oops, sorry! I forgot to update the image tags.

Thank you for telling me about this!

**UPDATE:** updated :)

**Comment by Alexander01998** [comment]
I get the following error when trying to run data migration:

```
--rpc-eval : RPC failed with reason :nodedown
```

I'm updating from v1.5.1, running on Ubuntu 22.04.2 LTS.

I saw a similar discussion in https://github.com/plausible/analytics/discussions/2896#discussioncomment-5986941 where the issue was caused by a `RELEASE_DISTRIBUTION=none` variable in the `plausible-conf.env` file. However, I don't have this variable in my env file and I still get the error. 😕

Any ideas?

**Reply by ruslandoga** [reply]
👋 @Alexander01998 

Could you please share your config files (docker compose, env) and describe your general setup (do you use plausible/hosting, what host are you using, etc.)?

**Reply by Alexander01998** [reply]
Hi @ruslandoga. Thanks for the quick response!

Yes, I use plausible/hosting. I also use nginx as my reverse-proxy, no mail server, and the default DB-IP database for IP geolocation.

My `docker-compose.yml` is as follows:

```yaml
version: "3.3"
services:
  mail:
    image: bytemark/smtp
    restart: always

  plausible_db:
    # supported versions are 12, 13, and 14
    image: postgres:14-alpine
    restart: always
    volumes:
      - db-data:/var/lib/postgresql/data
    environment:
      - POSTGRES_PASSWORD=redacted

  plausible_events_db:
    image: clickhouse/clickhouse-server:22.6-alpine
    restart: always
    volumes:
      - event-data:/var/lib/clickhouse
      - ./clickhouse/clickhouse-config.xml:/etc/clickhouse-server/config.d/logging.xml:ro
      - ./clickhouse/clickhouse-user-config.xml:/etc/clickhouse-server/users.d/logging.xml:ro
    ulimits:
      nofile:
        soft: 262144
        hard: 262144

  plausible:
    image: plausible/analytics:v1.5.1
    restart: always
    command: sh -c "sleep 10 && /entrypoint.sh db createdb && /entrypoint.sh db migrate && /entrypoint.sh db init-admin && /entrypoint.sh run"
    depends_on:
      - plausible_db
      - plausible_events_db
      - mail
    ports:
      - 127.0.0.1:8000:8000
    env_file:
      - plausible-conf.env

volumes:
  db-data:
    driver: local
  event-data:
    driver: local
  geoip:
    driver: local
```

My `plausible-conf.env` is as follows:

```properties
ADMIN_USER_EMAIL=redacted@example.com
ADMIN_USER_NAME=redacted
ADMIN_USER_PWD=redacted
BASE_URL=https://redacted.example.com
SECRET_KEY_BASE=redacted
LISTEN_IP=0.0.0.0
PORT=8000
DISABLE_REGISTRATION=true
GOOGLE_CLIENT_ID=redacted.apps.googleusercontent.com
GOOGLE_CLIENT_SECRET=redacted
```

Here is a timeline of how I tried to update to v2.0.0-rc.2:
- created a snapshot of the whole server
- `cd hosting`
- changed `image: plausible/analytics:v1.5.1` to `v2.0.0-rc.2` in `docker-compose.yml`
- stopped Plausible with `sudo docker-compose -f ~/hosting/docker-compose.yml down --remove-orphans`
- restarted Plausible with `sudo docker-compose -f ~/hosting/docker-compose.yml up -d`
- attempted migration with `sudo docker compose exec plausible bin/plausible rpc Plausible.DataMigration.NumericIDs.run`
- ran into the `--rpc-eval : RPC failed with reason :nodedown` error
- tried to run migration again, no error this time
- during migration, I was asked to confirm to drop a bunch of tables, which I did
- looked at the Plausible dashboard, nothing worked
- looked at the database with `docker compose exec plausible_events_db clickhouse-client`, `\c plausible_events_db`, `show tables;`
- saw `events` and `sessions`, but no `events_v2` or `sessions_v2`
- stopped and restarted Plausible again
- tried to load the dashboard again
- tried to run migration again
- gave up and restored the snapshot

Looking back at the timeline, I now realize that:
- I should have run `sudo docker compose rm plausible` before restarting plausible.
- I should have looked at the dashboard before starting data migration.
- I used a mix of `docker-compose` and `docker compose` without realizing the difference.

I'll try again to see if that fixes it. Please let me know if you spot any other issues with my setup.

**Reply by ruslandoga** [reply]
> ran into the --rpc-eval : RPC failed with reason :nodedown error

If you were executing these commands fast, then this error might have been due to `sleep 10` in `command` at [docker-compose.yml#L31](https://github.com/plausible/hosting/blob/dd0e109a57068cd47bae3c522260b1f971429650/docker-compose.yml#L31) Meaning, the node hadn't been started by that time yet.

> during migration, I was asked to confirm to drop a bunch of tables, which I did

Were you asked to confirm creation of any tables? :)

Full log of the data migration flow would be helpful.

> saw events and sessions, but no events_v2 or sessions_v2

Seems like the data migration didn't finish.

**Reply by Alexander01998** [reply]
Thanks for your help @ruslandoga. I've figured out what went wrong.

When I was running the data migration script for the first time, I entered `y` every time that it asked me to confirm something.

I figured this would be reasonable, as for example the `docker compose rm plausible` command also asks a `[y/N]` question and needs a `y` answer to work. But I now realize that the first two queries of the data migration script (`drop-events-v2` and `drop-sessions-v2`) are supposed to be denied in order for the script to run properly.

I've now re-run the script and simply pressed enter on every prompt so that it would run with all of the default yes/no answers. This worked. Everything has now been migrated successfully ([log output](https://gist.github.com/Alexander01998/600fdb6c4015c13cd721ce7e64b05c13)).

It would be great if the migration guide could be updated with a note about these confirmation prompts, as I'm sure I'm not the only one who would incorrectly assume that they should be answered with `y`.

**Reply by ruslandoga** [reply]
> docker compose rm plausible command also asks a [y/N] question and needs a y answer to work

I might be missing the point, but isn't it the same with the data migration script?

Entering `y<enter>` for `DROP TABLE ... [y/N]` drops the tables, `<enter>` or `n<enter>` -- it doesn't.

> the first two queries of the data migration script (drop-events-v2 and drop-sessions-v2) are supposed to be denied in order for the script to run properly

I'm not sure that's true, I've tried it both ways `y` / `n` (drop, no drop) and it worked both times, since these dropped tables are created at a later step: https://gist.github.com/Alexander01998/600fdb6c4015c13cd721ce7e64b05c13#file-plausible_data_migration-log-L50-L89, -- if the tables already exist, that step is no-op, if the tables have been dropped, they are created anew.

> t would be great if the migration guide could be updated with a note about these confirmation prompts, as I'm sure I'm not the only one who would incorrectly assume that they should be answered with y.

What should we write there?

**Reply by Alexander01998** [reply]
Huh, interesting. Every time I tried it with `y`, it didn't work. I'm pretty sure it never created the tables when I did that.

Since I still have my snapshot, let me run it again and get a log output for you...

**Reply by Alexander01998** [reply]
I can't reproduce the issue anymore. It just works now. https://gist.github.com/Alexander01998/eb9d5ffac55edf81025cd39ad1f391a4

**Comment by sgbx** [comment]
Sorry for the stupid question, I do not have much clue of docker. What/where exactly is the "hosting" folder?

**Reply by ruslandoga** [reply]
@pagenstedt no worries! The initial setup is described in https://plausible.io/docs/self-hosting

The hosting folder is just what I call the folder where the self hosting repo is cloned: `git clone https://github.com/plausible/hosting`

**Reply by sgbx** [reply]
OK, I see. Thank you.

**Reply by sgbx** [reply]
It was another issue. Docker Compose plugin was not installed. Now everything went fine.

**Comment by Eyesonly88** [comment]
What is the recommended and minimum CPU and RAM for this 2.0 version?

**Reply by ruslandoga** [reply]
👋 @Eyesonly88

I think it would be the same as before, https://plausible.io/docs/self-hosting#requirements :)

**Comment by ruslandoga** [comment]
`v2.0.0` has been released: https://github.com/plausible/analytics/discussions/3132

There are no significant changes from `v2.0.0-rc.2` so it should be safe to update.

**Reply by ruslandoga** [reply]
Changes: https://github.com/plausible/analytics/compare/v2.0.0-rc.2...v2.0.0

**Comment by chunkydotdev** [comment]
Upgraded using the instructions from [the docs](https://plausible.io/docs/self-hosting):
```
$ docker-compose down --remove-orphans
$ docker-compose pull plausible
$ docker-compose up -d
```
And thought I had lost all data, I really had a hallelujah-moment when I found this page! 😅

Please add the migration to the upgrade instructions-page so other people don't think they've deleted all their data after an upgrade. 🙏

**Reply by ruslandoga** [reply]
@Storken 👋

Right, the docs need to be updated. Maybe we can just link the discussions from there since they contain the most up to date info.

Thank you!

**Reply by ruslandoga** [reply]
I've added this note in https://github.com/plausible/docs/pull/399
